sklearn/utils/estimator_checks.py

```diff
@@ -103,7 +104,7 @@ def _yield_non_meta_checks(name, estimator):
         # cross-decomposition's "transform" returns X and Y
         yield check_pipeline_consistency
 
-    if name not in ALLOW_NAN:
+    if name not in ALLOW_NAN and not isinstance(estimator, SelectorMixin):
```

Comment:
Hmm... Not sure if we'd rather this. There still may be selectors (not in our library) that should error on NaN/inf, and many of these selectors with their default parameters still should/will error on NaN/inf.

In the (near) future, estimator tags (#8022) will solve this.

Reply:
Why do they fail the pickle test with default parameters? I think for default parameters all the feature selectors should still raise error on nan/inf except for VarianceThreshold so you should not otherwise need to modify estimator_checks should you?

Reply:
If you add them to ALLOW_NAN then they fail the pickle test because they don’t allow nan/inf with the default scorer. However if you don’t add them to ALLOW_NAN then check_estimators_nan_inf errors because the transform method does not check for nan/inf.

Reply:
Thanks for explaining.
I'm trying to work out if there's a better solution than creating `ALLOW_NAN_TRANSFORM`. Basically, at the moment we need to classify an estimator (and its parameter setting) as allowing or disallowing NaNs. Alternatively we could weaken the pickle test to fit on non-NaN data in the case that fitting with NaNs raises an error. (The ability to fit on NaN data there is not the point of the check. It's the interpretation of NaNs in `transform`/`predict` if NaNs are pickled with the model parameters/attributes that is the concern of that check.) Then we would be redefining ALLOW_NAN to mean "allows NaN in transform/predict, and may also allow NaN in fit". Your thoughts?

Reply:
Sorry this is harder than I thought!

Reply:
Thanks for helping work through it. I think relaxing the pickle test makes sense because it's not supposed to be directly checking NaN/Inf handling in the estimators. There are really two issues here: 
1) a discrepancy in whether an estimator allows NaN/Inf between its different methods (e.g. VarianceThreshold currently allows in transform but not fit), and 
2) the default parameter instance of an estimator not exhibiting its allowance of NaN/Inf. 

Your suggestion of `ALLOW_NAN_TRANSFORM` would solve 1 and I could try to implement it if we decide it's the right way to go. But 2 would still not be solved since these estimators would belong in the `ALLOW_NAN_FIT` category but still error in the pickle check with NaN. So I think weakening the pickle check is the best option without changing the default scorer. I've updated my code to include this new version of `check_estimators_pickle`.